### PR TITLE
Origin/w5/orm

### PR DESCRIPTION
**Pull request by Ghmeding**
A lot of changes.

Testing done and completed with both of the test files in /tests
As well as manual testing with the frontend and Postman for the API.

A next thing might be concurrency. 
That might be why our app crashes all the time.

### CODE REVIEW

File: go-minitwit/main.go [hunk 1]
```diff
@@ -132,31 +118,38 @@ func gravatarURL(email string, size int) string {
...
 		}
-		if email, ok := m["email"].(string); ok {
-			msg.Email = email
+		if reflect.TypeOf(m.Email).Kind() == reflect.String {
+			msg.Email = m.Email
 		}
-		if pubDate, ok := m["pub_date"].(int64); ok {
-			pubDateTime := time.Unix(pubDate, 0)
+		if reflect.TypeOf(m.PubDate).Kind() == reflect.Int {
+			pubDateTime := time.Unix(int64(m.PubDate), 0)
```
Comment (dudustri) [inline go-minitwit/main.go:152]: I think you can use normal int here

File: go-minitwit/main.go [hunk 2]
```diff
@@ -132,31 +118,38 @@ func gravatarURL(email string, size int) string {
...
 		}
-		if authorID, ok := m["author_id"].(int64); ok {
-			msg.AuthorID = int(authorID)
+		if reflect.TypeOf(m.AuthorID).Kind() == reflect.Int {
+			msg.AuthorID = int(m.AuthorID)
 		}
-		if userID, ok := m["user_id"].(int64); ok {
-			msg.User.UserID = int(userID)
+		if reflect.TypeOf(m.UserID).Kind() == reflect.Int {
+			msg.User.UserID = int(m.UserID)
```
Comment (dudustri) [inline go-minitwit/main.go]: why are you typecasting to int here? the ORM will not set it as int already since it reflects an integer?

File: go-minitwit/main.go [hunk 3]
```diff
@@ -117,8 +103,8 @@ func bytesToString(bytes [16]byte) string {
 }
 
 func checkPasswordHash(userEnteredPwd string, dbpwd string) bool {
-	bytes := md5.Sum([]byte(userEnteredPwd))
-	str := bytesToString(bytes)
+	hash := md5.Sum([]byte(userEnteredPwd))
```
Comment (dudustri) [inline go-minitwit/main.go:106]: can you explain why did you change it?

File: go-minitwit/db_methods.go [hunk 4]
```diff
@@ -230,76 +218,129 @@ func getUserByUsername(userName string) ([]map[string]interface{}, error) {
...
+	return nil
+
 }
 
 // adds a new message to the database
 func addMessage(text string, author_id string) error {
-	query := `insert into message (author_id, text, pub_date, flagged) values (?, ?, ?, 0)`
-	var db, err = connect_db(DATABASE)
+
+	authorIDintValue, err := strconv.Atoi(author_id)
```
Comment (dudustri) [inline go-minitwit/db_methods.go]: Here I would change author_id to int since we are converting to String to use the function and then converting back again to String. I think we should review all the usage of strconv Atoi and Itoa and define structures and functions where we don't need to do it. Probably we will increase a lot the performance. We don't need to do it in this PR but it is something to be tracked as a future task.

File: go-minitwit/api_handlers.go [hunk 5]
```diff
@@ -296,7 +296,7 @@ func apiMsgsPerUserHandler(c *gin.Context) {
 			c.AbortWithStatusJSON(http.StatusBadRequest, errorData)
 		}
 
-		err = addMessage(text, strconv.Itoa(int(authorId)))
+		err = addMessage(text, strconv.FormatInt(authorId, 10))
```
Comment (dudustri) [inline go-minitwit/api_handlers.go]: See addMessage comment.

File: go-minitwit/db_methods.go [hunk 6]
```diff
@@ -230,76 +218,129 @@ func getUserByUsername(userName string) ([]map[string]interface{}, error) {
...
+
 }
 
 // followUser adds a new follower to the database
 func followUser(userID string, profileUserID string) error {
-	query := `insert into follower (who_id, whom_id) values (?, ?)`
-	var db, err = connect_db(DATABASE)
-	if err != nil {
-		return err
+	userIDInt, errz := strconv.Atoi(userID)
```
Comment (dudustri) [inline go-minitwit/db_methods.go:266]: Same thing, maybe we can change it to int.

File: go-minitwit/db_methods.go [hunk 7]
```diff
@@ -230,76 +218,129 @@ func getUserByUsername(userName string) ([]map[string]interface{}, error) {
...
 }
 
 // unfollowUser removes a follower from the database
 func unfollowUser(userID string, profileUserID string) error {
-	fmt.Println(userID, profileUserID)
-	query := `delete from follower where who_id=? and whom_id=?`
-	var db, err = connect_db(DATABASE)
-	if err != nil {
-		return err
+	userIDInt, errz := strconv.Atoi(userID)
```
Comment (dudustri) [inline go-minitwit/db_methods.go:294]: Again here

File: go-minitwit/main.go [hunk 8]
```diff
@@ -171,23 +164,21 @@ func formatMessages(messages []map[string]interface{}) []Message {
...
-			filteredMsg.Content = text
+		if reflect.TypeOf(m.Text).Kind() == reflect.String {
+			filteredMsg.Content = m.Text
 		}
 
 		// publication date
-		if pubDate, ok := m["pub_date"].(int64); ok {
-			filteredMsg.PubDate = pubDate
-		}
+		filteredMsg.PubDate = int64(m.PubDate)
```
Comment (dudustri) [inline go-minitwit/main.go:177]: try only int instead of int64

File: go-minitwit/user_handlers.go [hunk 9]
```diff
@@ -80,13 +79,14 @@ func publicTimelineHandler(c *gin.Context) {
 }
 
 func userTimelineHandler(c *gin.Context) {
+	fmt.Println("userTimelineHandler")
```
Comment (dudustri) [inline go-minitwit/user_handlers.go]: remove

File: go-minitwit/user_handlers.go [hunk 10]
```diff
@@ -97,18 +97,23 @@ func userTimelineHandler(c *gin.Context) {
 
 	// does the logged in user follow them
 	followed := false
-	pUserId := profileUser[0]["user_id"].(int64)
-	profileName := profileUser[0]["username"]
+	pUserId := int64(profileUser.UserID)
```
Comment (dudustri) [inline go-minitwit/user_handlers.go]: I think UserID is already an int, isn't it?

Comment (dudustri) [reply go-minitwit/user_handlers.go:100]: also use normal int (less bytes = efficiency)

**Review by dudustri — CHANGES_REQUESTED** [review]
Really nice job. Just a few things xoxo

File: go-minitwit/main.go (hunk 1 above)
Comment (Ghmeding) [reply go-minitwit/main.go:152]: No. time.Unix() takes an int64

Comment (Ghmeding) [reply go-minitwit/main.go:152]: :)))))))))))))

File: go-minitwit/main.go (hunk 2 above)
Comment (Ghmeding) [reply go-minitwit/main.go]: noted and removed

File: go-minitwit/main.go (hunk 1 above)
Comment (dudustri) [reply go-minitwit/main.go:152]: alright, mb. But I think you are casting it to int64 when inserting in the PubDate field, no? then you don't need to type cast it again I guess.

File: go-minitwit/main.go (hunk 8 above)
Comment (dudustri) [reply go-minitwit/main.go:177]: forget about the previous comment

File: go-minitwit/main.go (hunk 3 above)
Comment (Ghmeding) [reply go-minitwit/main.go:106]: Okay. Long explanation. Before our hashing of password was a bit incorrect in terms of the data type being passed to the db. 

So, in the new registerUser() (db-method), I convert the hashed password into a string to get the proper format for the db.

This gave another bug, because now the formats where messed up for the checkPasswordHash() and i had to change that. But yeah, two lines so...

old one (buggy):
{32,44,185,98,172,89,7,91,150,75,7,21,45,35,75,112} ==  202cb962ac59075b964b07152d234b70

new one:

202cb962ac59075b964b07152d234b70  ==  202cb962ac59075b964b07152d234b70

Comment (dudustri) [reply go-minitwit/main.go:106]: amazing, well done

File: go-minitwit/main.go (hunk 1 above)
Comment (Ghmeding) [reply go-minitwit/main.go:152]: I am making big chances so we dont have so many type conversions.
The reason they are there was because of an old bug i think.

Comment (Ghmeding) [reply go-minitwit/main.go:152]: Will make a new PR

File: go-minitwit/db_methods.go (hunk 4 above)
Comment (Ghmeding) [reply go-minitwit/db_methods.go]: changing...

File: go-minitwit/api_handlers.go (hunk 5 above)
Comment (Ghmeding) [reply go-minitwit/api_handlers.go]: changing..

File: go-minitwit/db_methods.go (hunk 6 above)
Comment (Ghmeding) [reply go-minitwit/db_methods.go:266]: changing,..

File: go-minitwit/db_methods.go (hunk 7 above)
Comment (Ghmeding) [reply go-minitwit/db_methods.go:294]: ok changing

File: go-minitwit/user_handlers.go (hunk 9 above)
Comment (Ghmeding) [reply go-minitwit/user_handlers.go]: ok

File: go-minitwit/user_handlers.go (hunk 10 above)
Comment (Ghmeding) [reply go-minitwit/user_handlers.go]: ok changing

File: go-minitwit/main.go (hunk 1 above)
Comment (dudustri) [reply go-minitwit/main.go:152]: you dont need to do a new PR, just push it to the branch and this PR will be updated
